jax/_src/random.py

```diff
@@ -958,7 +958,13 @@ def _bernoulli(key, p, shape) -> Array:
   else:
     _check_shape("bernoulli", shape, np.shape(p))
 
-  return uniform(key, shape, lax.dtype(p)) < p
+  # we could return uniform(key, shape) < p. But uniform sacrifices some
+  # resolution, so instead we sample in the space of the generated bits.
+  nbits = dtypes.finfo(p.dtype).bits
+  unsigned_dtype = UINT_DTYPES[nbits]
```

Comment:
Why not just use always use uint64 here? (or 64 if np.finfo(np.float32).max > 2**64, 32 if ..., etc) ?

Reply:
Because the X64 flag prevents that in general

Reply:
ok, so uint32 instead? Also, to get the full uint64 bits worth of precision I think you can just do:

a = (p * 2**32).astype(uint32)
b = ((p * 2**32) - a).astype(uint32)
return select(bits0 == a, bits1 >= b, bits0 >= a)

Reply:
What is `bits0` and `bits1`?

Reply:
I think it would just be: `bits0, bits1 = bits(key, (2,) + shape, np.uint32)`

Reply:
Ah, I see. I was planning to punt on high-res mode for now (see PR description).

Reply:
I don't think we should double-sample by default, because most users don't need a full 64 bits.